### DISCUSSION TITLE
Feature Request: iOS Shortcuts

### DISCUSSION THREAD

**Discussion by acminor** [blinksh/blink#1592] [Ideas]
It would be nice to implement iOS shortcut functionality. This would (I assume) allow starting a connection without retyping or searching through history for the command.

I know because iOS restrictions, you probably cannot implement the running of shell scripts. So, this may be an alternative. Currently, I am using ed to create a file with the command which I can cat and retype or copy and paste if I cannot find the command with reverse-isearch.

**Comment by cloin** [comment]
a-shell does a nice job of this https://holzschu.github.io/a-Shell_iOS/

**Comment by ariez-xyz** [comment]
I would love a shortcut that takes a string and executes it as a command in Blink. I need to scp images a lot and that would really improve this workflow.

**Comment by christippett** [comment]
While native integration with Shortcuts would be nice, you can achieve similar functionality by calling Blink’s X-Callback-URL from within a Shortcut, e.g. `blinkshell://run?key=<URLKey>&cmd=<URLEncodedCommand>` (don’t forget to URL encode the command string).

I’m using exactly this method to send a wake-on-lan command to a device selected in an earlier Shortcuts step. Works great!

![Send WoL](https://user-images.githubusercontent.com/7218120/189320062-1b8da2e5-b91f-43fb-b620-c92061b1d866.png)

**Comment by iilyak** [comment]
@christippett Nice hack!!! Do you know any way to get output or status of the command?

**Comment by iilyak** [comment]
I am curious if there is a way to integrate [a-Shell_iOS/](https://holzschu.github.io/a-Shell_iOS/) and blink. I want to use shortcuts to run script in a-shell which would prepare some files and then I want to trigger blink command to upload these files to the server. The reason is:

- blink cannot run scripts
- a-shell don't have rsync or scp

**Comment by christippett** [comment]
> @christippett Nice hack!!! Do you know any way to get output or status of the command?

The built-in Open X-Callback URL action caters for that I'm pretty sure. I'm not using the output for anything in my example, but you can see there's an option to to parse a success or failed operation.

![image](https://user-images.githubusercontent.com/7218120/194224833-ef56a811-4a53-467b-bafa-6efdc45c264a.jpeg)

```
Open X-Callback URL

Performs the specified x-callback-url action. The x-success, x-cancel and x-error parameters will be added automatically. 
```

**Comment by carloscabanero** [comment]
As it has been mentioned already, although we do not have support for Shortcuts app, we do provide a X-Callback URL that becomes very handy. This plus piping can work on a few very cool scenarios, that I wanted to do a short video about. Here are some notes and additional tricks:
- You can execute snippets stored on Blink in the remote. Ie, you can create an `update.sh` script in Blink, and run it with `ssh host_name < update.sh`
- Using Shortcuts, I guess you could create a script that goes through a server list and runs update command.
- It is not possible to obtain the output of the command, or of stdout or stderr. X-Success URL is supported, so you could possible launch another command or another app afterwards.
- To access files from another app, you can use the link-files command, or use your iCloud folder as exchange. Link-files requires manual intervention on iOS, so probably using a different exchange folder may be better for this scenario.
- When downloading / uploading files multiple times between machines, it is possible to use a flag to preserve the metadata and only copy on change. You can use that to only scp the files that have been touched. 

I'm going to move this to Discussions, as there are a lot of tips here.

As per the feature request itself, we will see what we can do to put this in the roadmap sooner, and make things easier through better Shortcuts hooks.

**Comment by carloscabanero** [comment]
I'm reopening the conversation here as we have been thinking what use cases we could cover with this. Because we now have Snips, it would be a simple way to write a Snip and then execute it from a shortcut. This would cover basic scenarios like "regularly copying files from a remote". I was wondering what other use cases may be interesting for this.

Tailscale does a cool job with shortcuts for some of that. https://tailscale.com/kb/1233/mac-ios-shortcuts/

**Comment by iilyak** [comment]
Here is one use case. Quite often I receive a pdf which I want to store on my samba NAS. I want to implement a shortcut which would

1. ask me about the pdf: `subject`, `student`, `kind`, `grade`
2. add prefix to filename `{YYYY}-{MM}-{DD} - {subject} - {student} - {kind} - {seq}-{original-name}.pdf`
3. store this document on nas in `/school/{subject}-{student}/` folder

The second case is adding page numbers to pdf before printing. In order to do it I would need to run some linux tools. The pdf might be pre-processed already for double-sided printing in a booklet order, so achieving this functionality in shortcuts app is not possible.

**Reply by carloscabanero** [reply]
Thanks! The first scenario is great, we are trying to figure out more like this. PDF manipulation is a bit out of our scope, as you pointed out it would require other linux tools and we want to keep our base system small.

**Comment by iilyak** [comment]
> Thanks! The first scenario is great, we are trying to figure out more like this. PDF manipulation is a bit out of our scope, as you pointed out it would require other linux tools and we want to keep our base system small.

The reason I provided the second scenario is to give a use case where improved integration (via shortcuts mechanism) with [a-Shell_iOS/](https://holzschu.github.io/a-Shell_iOS/) might do the trick.

**Reply by carloscabanero** [reply]
That does make a lot of sense. Thanks!

**Comment by carloscabanero** [comment]
Linking here for more examples: https://github.com/blinksh/blink/issues/804

**Comment by cmartindmd** [comment]
Data point from the other side of this: **Blink 18.7.0 removed x-callback-url**, so the `blinkshell://run?key=…&cmd=…` workaround recommended in this thread no longer works.

From the 18.7.0 release notes: *"mosh1, ssh2 and x-callback-url have been removed"* and *"A new Commands system is on the way to replace x-callback-url; if that breaks a workflow you depend on, please get in touch."*

Worth flagging for anyone who lands here from a search: removing it also deregisters the `blinkshell://` scheme, so a Home Screen shortcut fails with **"Shortcuts could not open the app for the URL scheme 'blinkshell' because the app is not installed on this device."** That reads like a missing app and sends you hunting in the wrong place — and reinstalling Blink doesn't help, it just costs you the keys stored in the app.

My use case, in case it's useful while Commands is being designed:

iPhone → Tailscale → `mosh mac -- tmux-phone` into a persistent grouped tmux session on a Mac. One Home Screen tap put me in an attached session; that was the whole workflow.

What a replacement would need:

1. Triggerable from a Home Screen icon or a Shortcuts action, not only from inside Blink.
2. Runs the command and leaves you **attached and interactive** — landing in the session is the point, not returning output.
3. Can target a saved host.

No output capture needed (tmux holds the state), which makes this a narrower ask than the x-success chaining discussed upthread.

**Reply by christippett** [reply]
Argh, how disappointing. I created a new Shortcut just last week that used the `blinkshell://` URL scheme.

**Reply by carloscabanero** [reply]
This is a basic case and it will be covered very soon. I hope two weeks max will push first 18.8 version and at least this case will be there for sure. Plus, there will be more convenient ways later on with the new UI.

**Reply by christippett** [reply]
> This is a basic case and it will be covered very soon. I hope two weeks max will push first 18.8 version and at least this case will be there for sure. Plus, there will be more convenient ways later on with the new UI.

Thanks for providing a timeline for the update @carloscabanero. Apologies for the terseness of my previous reply, I look forward to seeing the new functionality!
